[FX] Lint pass for Graphs

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45227 [FX][WIP] Mutable Graph APIs
* #45245 [FX][WIP] Make return a node
* **#44973 [FX] Lint pass for Graphs**
* #45311 [FX][EZ] Fix bug where copying node made non-unique name

As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

Commit:
[FX] Lint pass for Graphs

[ghstack-poisoned]

Commit:
Update on "[FX] Lint pass for Graphs"


Would probably be useful for people debugging passes etc

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -243,10 +250,54 @@ def format_node(n : Node) -> Optional[str]:
             if node_str:
                 s += '\n    ' + node_str
-        if self.result:
+        if hasattr(self, 'result'):
```
hm, why hasattr? Why not to make it just None in the constructor?

jamesr66a replied:
idk, this is just how it was. We may be blasting this away in the future, so I'd prefer not to change it here

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -243,10 +250,54 @@ def format_node(n : Node) -> Optional[str]:
+
+        # Check targets are legit
+        for node in self.nodes:
+            if node.op in ['get_attr', 'call_module']:
```
do you want to check that `op` is valid too? (since it's just a string and someone could overwrite it later)

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -243,10 +250,54 @@ def format_node(n : Node) -> Optional[str]:
+            map_arg(self.result, check_arg)
+
+        # Check targets are legit
+        for node in self.nodes:
```
shall we check uniqueness of names too?

jamesr66a replied:
Yep, added that and it caught a bug https://github.com/pytorch/pytorch/pull/45311

Commit:
Update on "[FX] Lint pass for Graphs"


Would probably be useful for people debugging passes etc

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

Commit:
Update on "[FX] Lint pass for Graphs"


As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

zdevito reviewed (approved):
cool!

Commit:
Update on "[FX] Lint pass for Graphs"


As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

Commit:
Update on "[FX] Lint pass for Graphs"


As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

Commit:
Update on "[FX] Lint pass for Graphs"


As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

Commit:
Update on "[FX] Lint pass for Graphs"


As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

Commit:
Update on "[FX] Lint pass for Graphs"


As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]

Commit:
Update on "[FX] Lint pass for Graphs"


As we move to using mutable graphs, it's probably a good idea to have a lint pass to check that Graphs are well-formed

Differential Revision: [D23792631](https://our.internmc.facebook.com/intern/diff/D23792631)

[ghstack-poisoned]
